### DISCUSSION TITLE
想问下工具链以及工具链依赖的包的环境变量的作用域要怎么设置

### DISCUSSION THREAD

**Discussion by Yangff** [xmake-io/xmake#5171] [Q&A] [answered]
比如 `set_toolchains` 似乎只会在当前的xmake.lua作为主文件的时候生效，如果在另一个xmake.lua里include它，这个命令就失效了，哪怕相关的target和package是在之后引入的。。
比如
xmake.lua
```
include("RE-UE4SS")
```
RE-UE4SS/xmake.lua
```
set_toolchains(...)
include(targets...)
```

在根目录下xmake的时候所有的package都没有toolchain配置。不是很确定target里有没有生效，毕竟package都没编译完。。

另外就是，我试图
```
toolchain("zigcross")
    if is_host("windows") then
        set_toolset("cc", "$(scriptdir)/tools/zig/zig-cc.bat")
        set_toolset("cxx", "$(scriptdir)/tools/zig/zig-c++.bat")
        set_toolset("ld", "$(scriptdir)/tools/zig/zig-c++.bat")
        set_toolset("sh", "$(scriptdir)/tools/zig/zig-c++.bat")
        set_toolset("ar", "$(scriptdir)/tools/zig/zig-ar.bat")
    else
        set_toolset("cc", "$(scriptdir)/tools/zig/zig-cc")
        set_toolset("cxx", "$(scriptdir)/tools/zig/zig-c++")
        set_toolset("ld", "$(scriptdir)/tools/zig/zig-c++")
        set_toolset("sh", "$(scriptdir)/tools/zig/zig-c++")
        set_toolset("ar", "$(scriptdir)/tools/zig/zig-ar")
    end
    add_cxflags("-fexperimental-library")
    add_cxflags("-fno-delete-null-pointer-checks")
    add_cxflags("-gdwarf")
    add_cxflags("-fno-sanitize=undefined") -- can also use O2 to avoid this, but I'd prefer getting clear binary for now
    add_shflags("-z", "lazy")
toolchain_end()

add_requires("zig", {system = false})
set_toolchains("zigcross@zig", "rust")
```

这样来配置c/c++的工具链，package编译的时候它也会去用我设置的zig-cc，但是提示找不到zig，似乎是环境变量没有设置。
这个环境变量的作用域又是什么决定的，好像也没看到个说法。
（如果我没看错的话这样好像target是ok的，不过反正target我都可以手动加 `add_deps` 可能问题不大……吧……）

**Comment by waruqi** [comment, marked as answer]
> 在根目录下xmake的时候所有的package都没有toolchain配置。不是很确定target里有没有生效，毕竟package都没编译完。。

xmake 的所有 set_xxx 配置，都是按 includes tree 从顶层一层层往下继承的。。

你在最顶层 xmake.lua 的配置，所有子 xmake.lua 里面的 target 都会生效。。但是如果在某个 子 xmake.lua 里面设置 set_xxx，那么它平级的 xmake.lua 以及顶层 xmake.lua 里面的 target 是继承不到的，也就不会生效。

所以即使你在顶层 xmake.lua 去 includes 并里面 set_xxx 了，是不会生效的，你还是得在外面去设置。。如果设置比较多，可以搞个 init_xxx 的 function ，然后一起配置

xmake.lua

```lua
include("RE-UE4SS")

init_re_ue4ss()
```

RE-UE4SS/xmake.lua

```lua
function init_re_us4ss()
    set_toolchains(...)
    -- TODO
end

include(targets...)
```

**Comment by waruqi** [comment]
> 这样来配置c/c++的工具链，package编译的时候它也会去用我设置的zig-cc，但是提示找不到zig，似乎是环境变量没有设置。

内置的 zig toolchain 也是支持 zig cc 的，通常不需要自定义 toolchain

https://github.com/xmake-io/xmake/blob/master/tests/projects/c%2B%2B/console_zig_cxx/xmake.lua
https://github.com/xmake-io/xmake/blob/master/xmake/toolchains/zig/xmake.lua

**Reply by Yangff** [reply]
主要是我需要让它用我的.bat，不然1. 用了cmake的package不接受我的zig cc 2. rust的linker也不接受zig cc 它们都要求cc/ld指向一个文件

**Reply by Yangff** [reply]
而且默认的toolchain也是类似这样的定义吧……？
